text/0041-fixed-point.md

```diff
@@ -0,0 +1,104 @@
+- `.round(f_width=0)`: Return a new `FixedPointValue` with precision changed to `f_width`, rounding as required.
+- `.__add__(other)`, `.__radd__(other)`, `.__sub__(other)`, `.__rsub__(other)`, `.__mul__(other)`, `.__rmul__(other)`: Binary arithmetic operations.
+  - If `other` is a `Value` or an `int`, it'll be cast to a `FixedPointValue` first.
+  - If `other` is a `float`: TBD
+  - The result will be a new `FixedPointValue` with enough precision to hold any resulting value without rounding or overflowing.
+- `.__neg__()`, `.__pos__()`, `.__abs__()`: Unary arithmetic operations.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+TBD
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- `FixedPointValue.eq()` could cast a `Value` to a `FixedPointValue` first, and thereby shift the value to the integer part instead of assigning directly to the underlying value.
+  However, `Value.eq()` would always grab the underlying value of a `FixedPointValue`, and for consistency both `.eq()` operations should behave in the same manner.
+  - If we wanted to go the other way, this RFC could be deferred until another RFC allowing `ValueCastable` to override reflected `.eq()` have been merged.
+    However, having to explicitly do `value.eq(fp_value.round())` when rounding is desired is arguably preferable to having `value.eq(fp_value)` do an implicit rounding.
+
+- Unlike `.eq()`, it makes sense for arithmetic operations to cast a `Value` to `FixedPointValue`.
+  Multiplying an integer with a fixedpoint constant and rounding the result back to an integer is a reasonable and likely common thing to want to do.
+
+## Prior art
+[prior-art]: #prior-art
+
+[Q notation](https://en.wikipedia.org/wiki/Q_(number_format)) is a common and convenient way to specify floating point types.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What should we do if a `float` is passed as `other` to an arithmetic operation?
+  - We could use `float.as_integer_ratio()` to derive a perfect fixed point representation.
+    However, since a Python `float` is double precision, this means it's easy to make a >50 bit number by accident by doing something like `value * (1 / 3)`, and even if the result is rounded or truncated afterwards, the lower bits can affect rounding and thus won't be optimized out in synthesis.
+  - We could use the same width for `other` as for `self`, adjusted to the appropriate exponent for the value.
+  - We could outright reject it, requiring the user to explicitly specify precision like e.g. `value * Q(15).const(1 / 3)`.
+
+- There's two slightly different [Q notation](https://en.wikipedia.org/wiki/Q_(number_format)) definitions, namely whether the bit counts includes the sign bit or not.
+  `UQ(15)` and `UQ(7, 8)` would be 15 bits in either convention, but `Q(15)` and `Q(7, 8)` would be 15 or 16 bits depending on the convention. Which do we pick?
```

Comment:
In practice I'm not sure how common the "ARM" i.e. "include sign bit in n" convention is. Even Arm don't use it for modern software, e.g. "Q15" in [CMSIS-DSP](https://arm-software.github.io/CMSIS-DSP/latest/structarm__bilinear__interp__instance__q15.html). If we do have `Q(n)`, I think it should be `Q(15)` to mean a 16-bit signed value. However, if we did go with @whitequark's suggestion of `SQ(n)`, perhaps we have some more room to manoeuvre?

My preference would still be to omit the sign bit from the count, which presumably also matches the arguments to the `FixedPoint(f_width, /, *, signed)` constructor? In other words, I assumed `FixedPoint(15, signed=True)` creates a 16-bit `Value`, since it has 15 "fractional bits" which aren't sign bits?

Reply:
Yes, `Q(15)` or `SQ(15)` would be an alias for `FixedPoint(15, signed=True)` whichever way we decide to go.